eclair-core/src/main/scala/fr/acinq/eclair/payment/Bolt12Invoice.scala

```diff
@@ -39,57 +39,34 @@ case class Bolt12Invoice(records: TlvStream[InvoiceTlv]) extends Invoice {
-  override val relativeExpiry: FiniteDuration = FiniteDuration(records.get[RelativeExpiry].map(_.seconds).getOrElse(DEFAULT_EXPIRY_SECONDS), TimeUnit.SECONDS)
-  override val minFinalCltvExpiryDelta: CltvExpiryDelta = records.get[Cltv].map(_.minFinalCltvExpiry).getOrElse(DEFAULT_MIN_FINAL_EXPIRY_DELTA)
-  override val features: Features[InvoiceFeature] = records.get[FeaturesTlv].map(_.features.invoiceFeatures()).getOrElse(Features.empty)
-  val chain: ByteVector32 = records.get[Chain].map(_.hash).getOrElse(Block.LivenetGenesisBlock.hash)
-  val offerId: Option[ByteVector32] = records.get[OfferId].map(_.offerId)
-  val blindedPaths: Seq[RouteBlinding.BlindedRoute] = records.get[Paths].get.paths
-  val blindedPathsInfo: Seq[PaymentInfo] = records.get[PaymentPathsInfo].get.paymentInfo
-  val issuer: Option[String] = records.get[Issuer].map(_.issuer)
-  val quantity: Option[Long] = records.get[Quantity].map(_.quantity)
-  val refundFor: Option[ByteVector32] = records.get[RefundFor].map(_.refundedPaymentHash)
-  val payerKey: Option[ByteVector32] = records.get[PayerKey].map(_.publicKey)
-  val payerNote: Option[String] = records.get[PayerNote].map(_.note)
-  val payerInfo: Option[ByteVector] = records.get[PayerInfo].map(_.info)
-  val fallbacks: Option[Seq[FallbackAddress]] = records.get[Fallbacks].map(_.addresses)
-  val refundSignature: Option[ByteVector64] = records.get[RefundSignature].map(_.signature)
-  val replaceInvoice: Option[ByteVector32] = records.get[ReplaceInvoice].map(_.paymentHash)
+  override val nodeId: Crypto.PublicKey = records.get[InvoiceNodeId].get.nodeId
+  override val paymentHash: ByteVector32 = records.get[InvoicePaymentHash].get.hash
+  override val description: Either[String, ByteVector32] = Left(invoiceRequest.offer.description)
+  override val createdAt: TimestampSecond = records.get[InvoiceCreatedAt].get.timestamp
+  override val relativeExpiry: FiniteDuration = FiniteDuration(records.get[InvoiceRelativeExpiry].map(_.seconds).getOrElse(DEFAULT_EXPIRY_SECONDS), TimeUnit.SECONDS)
+  override val minFinalCltvExpiryDelta: CltvExpiryDelta = CltvExpiryDelta(0)
+  override val features: Features[InvoiceFeature] = records.get[InvoiceFeatures].map(_.features.invoiceFeatures()).getOrElse(Features.empty)
+  val blindedPaths: Seq[RouteBlinding.BlindedRoute] = records.get[InvoicePaths].get.paths
+  val blindedPathsInfo: Seq[PaymentInfo] = records.get[InvoiceBlindedPay].get.paymentInfo
+  val fallbacks: Option[Seq[FallbackAddress]] = records.get[InvoiceFallbacks].map(_.addresses)
   val signature: ByteVector64 = records.get[Signature].get.signature
 
   // It is assumed that the request is valid for this offer.
-  def isValidFor(offer: Offer, request: InvoiceRequest): Boolean = {
-    nodeId == offer.nodeId &&
-      checkSignature() &&
-      offerId.contains(request.offerId) &&
-      request.chain == chain &&
+  def isValidFor(request: InvoiceRequest): Boolean = {
```

Comment:
Is this function still useful now that the invoice request is embedded?

Reply:
It is a bit misleading though, we're not checking that the invoice is valid for a given invoice request (which implies it could be valid for various invoice requests), we want it to exactly match the invoice request we sent.

Wouldn't it make more sense to take an `invoiceRequest: InvoiceRequest` argument in `Bolt12Invoice.validate` and verify that it matches? It's quite confusing to have `Bolt12Invoice.validate` and `isValidFor`, especially since both must be called before actually using the invoice?

Reply:
For me these two steps are quite distinct:
- `validate` checks that the TLVS is well-formed and returns the corresponding case class, it doesn't check any business logic, just that it parses correctly. Actually that's where I think the name is misleading, I would call it something like `create`, `parse` or `decode`.
- `isValidFor` checks that the other party is not trying to cheat.

For instance `validate` is currently used in `Invoice.fromString` and I don't think it would make sense to require an invoice request just to parse a string, especially since this method is shared with `Bolt11Invoice`.

Reply:
That makes sense, thanks for the explanation!
I agree with you, we should rename the current `validate` functions of the Bolt 12 types.
I don't think `parse` and `decode` are a good match, the contents have already been parsed/decoded from binary.
These functions verify that the tlv stream contains all mandatory content, I guess `validateTlvs` is a better match?
This way it's a bit more obvious that we only validated tlvs, and further validation is needed afterwards to verify that the other party isn't cheating.